File: lib/services/moralis/moralis.dart
import 'dart:js_util';

import 'package:injectable/injectable.dart';
import 'moralis_interop.dart' as moralisJS;

@Singleton()
class Moralis {
  Future<void> login() async => await promiseToFuture<void>(moralisJS.login());

  Future<void> logout() async =>
      await promiseToFuture<void>(moralisJS.logout());

  Future<bool> isLoggedIn() async =>
      await promiseToFuture<bool>(moralisJS.isLoggedIn());

  Future<String> upload(String id, dynamic base64) async =>
      await promiseToFuture<String>(moralisJS.upload(id, base64));
}

File: web/moralis.js
const serverUrl = "https://pym3qnzp1aax.usemoralis.com:2053/server";
const appId = "fJAAtDS7hryZ3HqBseVUWpPD2tsrZWUBAZiikbaY";
console.log("Startting up Moralis");
Moralis.start({ serverUrl, appId });
console.log("Moralis Initialized");

async function login() {
  let user = Moralis.User.current();
  if (!user) {
    user = await Moralis.authenticate({
      signingMessage: "Sign this message in order to login.",
    }).then(function (user) {
      console.log("logged in user:", user);
      console.log(user.get("ethAddress"));
    });
  }
}

async function logout() {
  await Moralis.User.logOut();
  console.log("logged out");
}

async function isLoggedIn() {
  return Moralis.User != null;
}

async function upload(moduleId, rawJsonBase64) {
  const file = new Moralis.File(moduleId + ".json", { base64: rawJsonBase64 });
  await file.saveIPFS();
  return file.hash();
}
